### ISSUE TITLE
Unable to allocate memory error

### ISSUE THREAD

**Issue by reb197** [ing-bank/sparse_dot_topn#33]
I using this library via the TextPack package to detect duplicate company names. It has been fine for a few months but yesterday I got an out of memory error, "Unable to allocate 5.37 GiB for an array with shape (720117225,) and data type float64" at sparse_dot_topn/awesome_cossim_topn.py in awesome_cossim_topn, line 58. The code is:
```
data = np.empty(nnz_max, dtype=A.dtype)
```
The production environment is using a celery worker thread inside a docker container on AWS. If I restart the container and run it manually it works but fails again the next day.

There could be a memory leak (perhaps from another unrelated task) but I do not understand how this call to `np.empty` ever works - the task in the container only has 3584 MiB RAM, yet this call is requesting an array gigabytes in size. How come it doesn't crash each time?

**Comment by ymwdalex** [comment]
@reb197 thanks for your feedback!

About why `np.empty does not crash **immediately** when create a 5GB array`, I think it is related to the mechanism of calloc (the low level implementation of numpy.empty https://github.com/numpy/numpy/blob/590facafcbd454fc02f96aa52c1c3412d6cccd14/numpy/core/src/multiarray/alloc.c#L212) and linux kernel memory allocate mechanism. There are some discussions like [this](https://stackoverflow.com/questions/27574881/why-does-numpy-zeros-takes-up-little-space) and [this](https://stackoverflow.com/questions/51748215/why-is-my-numpy-array-taking-much-less-memory-than-it-should). You can also do the experiment that np.empty(2<<40) in your machine. You don't get immediately crashed.

In your running environment, is the repeatable crashing always after a 5.37 GiB array creation? If yes, then I think it makes sense. As you said, your celery worker doesn't get enough memory. The crashing is not happened immediately, but when sparse_dot_topn is filling the result matrix, OS will raise the memory error.

I think you can:
- either increase your memory size
- or preprocessing your input data set. maybe split it into multiple smaller data set, do the deduplication, then merge them again.

**Comment by reb197** [comment]
Thanks @ymwdalex, those links are very helpful.

I made Celery restart the worker process after each task (--max-tasks-per-child 1) and it ran OK. This _may_ have solved the problem but I want to see how it runs over the next days to be sure.

It crashed on the array allocation, not subsequent access, so Linux was denying the _calloc_ memory request. I have found that by default Linux uses heuristics (based on counting free memory pages and swap space) to decide when to allow requests for very large amounts of memory, known as an overcommit.

My thinking is that the code is requesting an amount very close to the upper limit for the machine – when few other pages of RAM are in use it works but fails otherwise. Restarting the worker process may be giving it the best chance of success.

The overcommit heuristics setting can be adjusted with `sysctl vm.overcommit_memory`
 such that overcommits are always allowed, but then there is a greater risk the kernel will kill processes or give a page fault when accessing memory. I might still try it.

It is not an issue with this project, so closing the issue.

Further reading: 

http://engineering.pivotal.io/post/virtual_memory_settings_in_linux_-_the_problem_with_overcommit/
https://www.kernel.org/doc/Documentation/vm/overcommit-accounting
https://stackoverflow.com/questions/38688824/linux-over-commit-heuristic
